File: animations-java/src/main/java/dev/hogoshi/animations/animation/Animation.java
package dev.hogoshi.animations.animation;

import dev.hogoshi.animations.easing.Easing;
import dev.hogoshi.animations.easing.Easings;

/**
 * Basic animation
 *
 * @author hogoshi
 */
public class Animation extends AbstractAnimation {

    public void animate(double valueTo, double duration) {
        animate(valueTo, duration, Easings.LINEAR);
    }

    public void animate(double valueTo, double duration, Easing easing) {
        animate(valueTo, duration, easing, false);
    }

    public void animate(double valueTo, double duration, Easing easing, boolean debug) {
        setDebug(debug);
        if (isAlive() && (valueTo == getFrom() || valueTo == getTo() || valueTo == getValue())) {
            if (isDebug()) System.out.println("Animate cancelled due to valueTo equals fromValue");
            return;
        }

        setEasing(easing);
        setDuration(duration * 1000);
        setStart(System.currentTimeMillis());
        setFrom(getValue());
        setTo(valueTo);

        if (isDebug())
            System.out.println("#animate {\n    to value: " + getTo() + "\n    from value: " + getValue() + "\n    duration: " + getDuration() + "\n}");
    }

}

File: animations-kotlin/src/main/kotlin/dev/hogoshi/animations/animation/Animation.kt
package dev.hogoshi.animations.animation

import dev.hogoshi.animations.easing.Easing
import dev.hogoshi.animations.easing.Easings

/**
 * Basic animation
 * @author hogoshi
 */
class Animation : AbstractAnimation() {

    fun animate(valueTo: Double, duration: Double) = animate(valueTo, duration, Easings.LINEAR)

    fun animate(valueTo: Double, duration: Double, easing: Easing) = animate(valueTo, duration, easing, false)

    fun animate(valueTo: Double, duration: Double, easing: Easing, debug: Boolean) {
        this.debug = debug
        if (isAlive() && (valueTo == from || valueTo == to || valueTo == value)) {
            if (debug) println("Animate cancelled due to valueTo equals fromValue")
            return
        }

        this.easing = easing
        this.duration = duration * 1000
        this.start = System.currentTimeMillis()
        this.from = value
        this.to = valueTo

        if (debug) println(
            """#animate {
                to value: $to
                from value: $value
                duration: $duration
            }"""
        )
    }

}